### PR TITLE
Adding vehicle rearming

### PR DESCRIPTION
**Pull request by GitHawk**
As discussed with @Glowbal on slack, a PR for including vehicle rearming into ACE.

TO DO
- [ ] Need someone to do translation

NICE TO HAVE
- [ ] More carryable models
- [ ] Limited ammo amount as module option
- [ ] Module to assign rearm vehicle (with options to limit ammo type and amount - read below)
- [ ] Module to set ammo type available (in the synchronized/inputted rearm vehicle) - requires expanding the framework
- [ ] Rearm facilities (location that lets you fully rearm a vehicle, without limits, would always rearm entire vehicle no matter the level)
- [ ] Module to assign rearm facility (requires above)
- [ ] Curator module to assign rearm vehicle (with default limits or custom selection, however that's done in with curator modules)
- [ ] Curator module to assign rearm facility

DONE
- [x] Core coding
- [x] Module init working
- [x] SP testing
- [x] MP testing
- [x] Module icon
- [x] Implement different rearm levels
- [x] Code cleanup
- [x] Fixed magazines without displayName
- [x] Fix some weird bugs
- [x] Fixed rearming driver magazines when they are depleted
- [x] Enable dropping of ammo
- [x] Add generic ammo dummy
- [x] Drop currently held ammo to the ground when entering a vehicle

### CODE REVIEW

File: addons/rearm/ACE_Settings.hpp [hunk 1]
```diff
@@ -0,0 +1,9 @@
+class ACE_Settings {
+    class GVAR(level) {
+        displayName = LSTRING(RearmSettings_level_DisplayName);
+        description = LSTRING(RearmSettings_level_Description);
+        value = 1;
+        typeName = "SCALAR";
+        values[] = {LSTRING(RearmSettings_vehicle), LSTRING(RearmSettings_magazine), LSTRING(RearmSettings_caliber)};
```
Comment (jonpas) [inline addons/rearm/ACE_Settings.hpp]: `CSTRING`

File: addons/rearm/ACE_Settings.hpp [hunk 2]
```diff
@@ -0,0 +1,9 @@
+class ACE_Settings {
+    class GVAR(level) {
+        displayName = LSTRING(RearmSettings_level_DisplayName);
```
Comment (jonpas) [inline addons/rearm/ACE_Settings.hpp]: `CSTRING`

File: addons/rearm/ACE_Settings.hpp [hunk 3]
```diff
@@ -0,0 +1,9 @@
...
+        description = LSTRING(RearmSettings_level_Description);
```
Comment (jonpas) [inline addons/rearm/ACE_Settings.hpp]: `CSTRING`

**Comment by alganthe** [comment]
No need to fix RHS issues for the first batch, you can do that later.

**Comment by Glowbal** [comment]
Could you give us a brief description of how this module works once you have it functional?

**Comment by GitHawk** [comment]
@alganthe I believe I found the bug since I could reproduce it with a vanilla vehicle. But since RHS uses multiple turrets and multiple magazines of the same type per turret, so I can test the code quite well.

@Glowbal Sure, it's functional. You have 3 different rearming "levels": 
- On the first level, you rearm an entire vehicle in 10 seconds by interacting with an ammo truck (replacement for vanilla rearm).

- On the second level you have to individually pick up a fresh magazine at the ammo truck and "carry" it to the vehicle where you have to put it in. Picking up ammo and rearming a vehicle takes some time. Fresh magazine means vanilla mags e.g. 5000Rnd_762x51_Belt or 40Rnd_105mm_APFSDS. As a player, you have to do some running around but you can rearm a vehicle in something like a minute or two.

- The third level is similar to the second, but you no longer carry an entire magazine. The number rounds you carry depends on the caliber. For example: you carry 400 rounds of 7.62x51mm and can rearm that 5000 round belt in up to 13 runs. Since the 105mm rounds are much bigger and heavier, you can only carry a single round and need to do 40 runs to rearm the tank's main gun. Now it takes a single player anything from a few seconds to over 6 minutes to rearm a single magazine (depending on caliber an rounds spent). The time needed may require some balancing or introduction of a speed factor in the module.

The pick up time, rearm time and caliber dependent round count are found in ``` addons/rearm/script_component.hpp```.

**Comment by GitHawk** [comment]
At least some rocket and bombs have useful models:
Rearming an A10-D with a Maverick

File: addons/rearm/CfgMagazines.hpp [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+    class VehicleMagazine;
+    class SmokeLauncherMag : VehicleMagazine {
+        displayName = CSTRING(smokelauncher);
+    };
+    class SmokeLauncherMag_boat : VehicleMagazine {
+        displayName = CSTRING(smokelauncher);
+    };
+    
+    class 1000Rnd_Gatling_30mm_Plane_CAS_01_F : VehicleMagazine {
+        displayName = "30mm HEI";
```
Comment (jonpas) [inline addons/rearm/CfgMagazines.hpp]: Those display names should be stringtabled, either here or in realisticnames.

**Comment by jonpas** [comment]
Look awesome!

File: addons/rearm/CfgMagazines.hpp (hunk 4 above)
Comment (GitHawk) [inline addons/rearm/CfgMagazines.hpp]: I know. Realistic names only changes the weapon names, not the magazine names. But I will use ECSTRING if they add it.

Comment (jonpas) [inline addons/rearm/CfgMagazines.hpp]: Cool, for now can you put them in a stringtable file? That way they can be translated properly where needed (and will be easier to move them if that will be decided).

**Comment by gienkov** [comment]
AGM-65K is 306 kg :>.

File: addons/rearm/CfgAmmo.hpp [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+        GVAR(caliber) = 250; // Default caliber for bombs
+        GVAR(dummy) = QGVAR(Bomb_04_F);
+    };
+    
+    class Bomb_03_F : Bomb_04_F {
+        GVAR(dummy) = QGVAR(Bomb_03_F);
+    };
+    
+    // F18 Mod
+    class js_a_fa18_wing_tank : LaserBombCore {
```
Comment (Glowbal) [inline addons/rearm/CfgAmmo.hpp]: These are not vanilla ammuniation. Should not be in the rearming pbo.

**Comment by GitHawk** [comment]
@gienkov If you make a model for something like that:
Ammo carrier

 I'll use it ;-)

File: addons/rearm/functions/fnc_getMaxMagazines.sqf [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+        _cfg = configFile >> "CfgVehicles" >> (typeOf _target) >> "Turrets";
+        if (count _cfg > 0) then {
+            _cfg = _cfg select 0;
+        } else {
+            _cfg = configFile >> "CfgVehicles" >> (typeOf _target);
+        };
+    };
+    case [1] : {
+         _cfg = configFile >> "CfgVehicles" >> (typeOf _target) >> "Turrets";
+        if (count _cfg > 0) then {
```
Comment (GitHawk) [inline addons/rearm/functions/fnc_getMaxMagazines.sqf]: Should be > 1

File: addons/rearm/functions/fnc_getMaxMagazines.sqf [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+            _cfg = _cfg select 1;
+        } else {
+            _cfg = configFile >> "CfgVehicles" >> (typeOf _target);
+        };
+    };
+    case [2] : {
+        _cfg = configFile >> "CfgVehicles" >> (typeOf _target) >> "Turrets";
+        if (count _cfg > 0) then {
```
Comment (GitHawk) [inline addons/rearm/functions/fnc_getMaxMagazines.sqf]: Should be > 2

File: addons/rearm/functions/fnc_getMaxMagazines.sqf [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+                _cfg = configFile >> "CfgVehicles" >> (typeOf _target);
+            };
+        } else {
+            _cfg = configFile >> "CfgVehicles" >> (typeOf _target);
+        };
+    };
+    case [0,1] : {
+        if (count _cfg > 0) then {
+            _cfg = (_cfg select 0) >> "Turrets";
+            if (count _cfg > 0) then {
```
Comment (GitHawk) [inline addons/rearm/functions/fnc_getMaxMagazines.sqf]: Should be > 1

File: addons/rearm/functions/fnc_getMaxMagazines.sqf [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+                _cfg = _cfg select 0;
+            } else {
+                _cfg = configFile >> "CfgVehicles" >> (typeOf _target);
+            };
+        } else {
+            _cfg = configFile >> "CfgVehicles" >> (typeOf _target);
+        };
+    };
+    case [0,1] : {
+        if (count _cfg > 0) then {
```
Comment (GitHawk) [inline addons/rearm/functions/fnc_getMaxMagazines.sqf]: _cfg = configFile >> "CfgVehicles" >> (typeOf _target) >> "Turrets";
is missing

**Comment by jonpas** [comment]
All comments resolved, bunch of fixes, improvements, changes in https://github.com/GitHawk/ACE3/pull/2

New things found for to-do list:
- [ ] Fix: Rearming flares when they are depleted
- [ ] Fix: [ACE] ERROR: Undefined Ammo [CMflare_Chaff_Ammo : bin\config.bin/CfgAmmo/CMflareAmmo]"
- [ ] Change: Prevent getting in when holding ammo

**Comment by GitHawk** [comment]
@jokoho48 Can you take a look at my last push?

**Comment by GitHawk** [comment]
@Glowbal @jokoho48 Needs a review and some independent testing. Version 1.0 is ready.

**Comment by Glowbal** [comment]
Thank you @GitHawk 

I will look at this sometime this weekend.

**Comment by GitHawk** [comment]
Heres a video of our clan testing it:
https://www.youtube.com/watch?v=LowPTMCjtjc

**Comment by GitHawk** [comment]
@Glowbal Still need reviewing. Our clan is succesfully using it since start of September.

**Comment by Glowbal** [comment]
Same as with #2141, I've handed this off to some of my testers. Pending feedback and review, this should be good to go in 3.4.0. Don't worry, I will sort out the merge conflicts on this one.

**Comment by MoistBaguette** [comment]
Tested it out for a bit and it looks pretty good.

So far, the only thing I noticed is that when you pick up an ammo box while crouched, you start to get pushed into the air as shown in the video below:
https://www.youtube.com/watch?v=wLlK67oERiU (At 0:22)

**Comment by GitHawk** [comment]
@MoistBaguette Looks like a collision problem. Will have a look into it.

**Comment by GitHawk** [comment]
@MoistBaguette Should no longer produce flying.

**Comment by MoistBaguette** [comment]
Thanks @GitHawk, works like a charm :thumbsup:

**Comment by alganthe** [comment]
You can use `local` (soon to be `private` to private local vars, it's faster and easier to read.

**Comment by GitHawk** [comment]
@Glowbal I did not get any additional feedback from your testers. The only bug they reported is fixed. We have been using it in missions with up to 50 players since September.
I hope it will make it into 3.4.

**Comment by Glowbal** [comment]
Thanks for that change @GitHawk.

**Comment by gienkov** [comment]
I've tested this on dedicated with some ppl and didn't saw a problem with it.

**Pull request merged**